### PR TITLE
Made Subfactorial iterative

### LINKED ISSUE #10962: maximum recursion depth exceeded on subfactorial(1000)

### PR DESCRIPTION
**Pull request by rd13123013**
Fixes #10962

### CODE REVIEW

File: sympy/functions/combinatorial/factorials.py [hunk 1]
```diff
@@ -210,9 +210,11 @@ class subfactorial(CombinatorialFunction):
         !n = {  0                             for n = 1
              |  (n - 1)*(!(n - 1) + !(n - 2)) for n > 1
               `
+    recurrence can be reduced to 
 
-    It can also be written as int(round(n!/exp(1))) but the recursive
-    definition with caching is implemented for this function.
+    !n = n*( !(n-1) ) + (-1)**n
```
Comment (leosartaj) [inline sympy/functions/combinatorial/factorials.py]: ``!n = n*(!(n-1)) + (-1)**n``

Comment (leosartaj) [inline sympy/functions/combinatorial/factorials.py]: You should also add a note, regarding which formula is used for computing subfactorial.

**Comment by leosartaj** [comment]
Can you comment on the performance difference, this formula introduces?

File: sympy/functions/combinatorial/factorials.py [hunk 2]
```diff
@@ -210,9 +210,12 @@ class subfactorial(CombinatorialFunction):
         !n = {  0                             for n = 1
              |  (n - 1)*(!(n - 1) + !(n - 2)) for n > 1
               `
+    recurrence can be reduced to
 
-    It can also be written as int(round(n!/exp(1))) but the recursive
-    definition with caching is implemented for this function.
+    !n = n*(!(n-1)) + (-1)**n
```
Comment (asmeurer) [inline sympy/functions/combinatorial/factorials.py]: You need to format this correctly for Sphinx.

**Comment by asmeurer** [comment]
It's already way faster. But it seems my method would be even better, as it avoids taking `(-1)**n`. You might want to time the two to check, though.

**Comment by czgdp1807** [comment]
The purpose of this PR has been fulfilled by #17983 
Therefore closing this for now. Thanks for the contributions. Please let us know if there is a need to re-open this.

**Pull request closed without merging**